### PR TITLE
many: allow installations for partial gadgets

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Allow installations for partial gadgets. muinstaller is modified to support this case.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
This is ready for an initial architectural review, after that my plan is to split this PR.

File: gadget/gadget.go [hunk 1]
```diff
@@ -256,22 +260,17 @@ func (v *VolumeStructure) IsRoleMBR() bool {
...
-// HasFilesystem returns true if the structure is using a filesystem.
-func (vs *VolumeStructure) HasFilesystem() bool {
-	return vs.Filesystem != "none" && vs.Filesystem != ""
-}
-
-// willHaveFilesystem considers also partially defined filesystem.
-func (vs *VolumeStructure) willHaveFilesystem(v *Volume) bool {
-	if vs.HasFilesystem() {
+// HasFilesystem returns true if the structure definition expects a filesystem.
+func (vs *VolumeStructure) HasFilesystem(v *Volume) bool {
```
Comment (pedronis) [inline gadget/gadget.go:264]: how annoying would be to introduce backpointers from VolumeStructure to Structure instead of this? it also make the change hard to follow atm, if we can't avoid this it needs to be done separately from using partial information as this is quite a churny change

File: gadget/layout.go [hunk 2]
```diff
@@ -120,9 +120,9 @@ func (l LaidOutStructure) Role() string {
 	return l.VolumeStructure.Role
 }
 
-// HasFilesystem returns true if the structure is using a filesystem.
-func (l *LaidOutStructure) HasFilesystem() bool {
-	return l.VolumeStructure.HasFilesystem()
+// HasFilesystem returns true if the gadget expects a filesystem.
+func (l *LaidOutStructure) HasFilesystem(lv *LaidOutVolume) bool {
```
Comment (pedronis) [inline gadget/layout.go:124]: same question

**Review by pedronis — COMMENTED** [review]
did a first pass but this a bit hard to follow as one change, questions about the need to pass in *Volume in places

File: gadget/gadget.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget.go:264]: Thinking about this, maybe the cleaniest will be to have a new variable in the structure that tells us if the structure requires a filesystem or not, as in the end that is implicit information encoded in gadget.yaml. The variable can then be set when reading this file.

Comment (pedronis) [reply gadget/gadget.go:264]: that sounds fine assuming we don't have other potential methods on *Structure that will need to know partial content? Seems it should be done as a preparatory PR before this one

**Comment by alfonsosanchezbeato** [comment]
PR rebased and history cleaned-up now

File: gadget/update.go [hunk 3]
```diff
@@ -336,13 +336,13 @@ func EnsureVolumeCompatibility(gadgetVolume *Volume, diskVolume *OnDiskVolume, o
...
 		opts = &VolumeCompatibilityOptions{}
 	}
-	logger.Debugf("checking volume compatibility between gadget volume %s and disk %s",
-		gadgetVolume.Name, diskVolume.Device)
+	logger.Debugf("checking volume compatibility between gadget volume %s (partial: %v) and disk %s",
+		gadgetVolume.Name, gadgetVolume.Partial, diskVolume.Device)
 
-	eq := func(ds *OnDiskStructure, vss []VolumeStructure, vssIdx int) (bool, string) {
-		gv := &vss[vssIdx]
+	eq := func(ds *OnDiskStructure, gv *Volume, vssIdx int) (bool, string) {
```
Comment (pedronis) [inline gadget/update.go:342]: do we need the the change of signature? and if we do, do we need gv to be passed in? as per my other comment

**Review by pedronis — COMMENTED** [review]
thanks, did a pass, did not fully look at update.go because there are some larger org questions there first, also didn't yet fully look at the spread test changes

File: overlord/devicestate/handlers_install.go [hunk 4]
```diff
@@ -961,6 +961,12 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 		return fmt.Errorf("on finish install: cannot layout volumes: %v", err)
 	}
 
+	// Import new information from the installer to the laid out data,
+	// so the gadget is not partially defined anymore if it was.
+	if err := gadget.ApplyInstallerVolumesToGadget(onVolumes, allLaidOutVols); err != nil {
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go:966]: I have added a test now

File: gadget/update.go [hunk 5]
```diff
@@ -530,23 +535,26 @@ func EnsureVolumeCompatibility(gadgetVolume *Volume, diskVolume *OnDiskVolume, o
...
-	for _, ds := range diskVolume.Structure {
-		present, reasonAbsent := gadgetContains(gadgetVolume.Structure, &ds)
-		if !present {
-			if reasonAbsent != "" {
-				// use the right format so that it can be
-				// appended to the error message
-				reasonAbsent = fmt.Sprintf(": %s", reasonAbsent)
+	// (unless partial strucuture).
+	if !gadgetVolume.HasPartial(PartialStructure) {
+		for _, ds := range diskVolume.Structure {
```
Comment (pedronis) [inline gadget/update.go:548]: shouldn't we still check that all the structures in the gadget exist on disk? before there was a one-to-one mapping but it seems we simply don't check anything now if structure is partial

File: gadget/update.go [hunk 6]
```diff
@@ -1474,9 +1495,14 @@ func canUpdateStructure(fromVss []VolumeStructure, fromIdx int, toVss []VolumeSt
...
-			return fmt.Errorf("cannot change filesystem from %q to %q",
-				from.Filesystem, to.Filesystem)
+		if !fromV.HasPartial(PartialFilesystem) {
+			if toV.HasPartial(PartialFilesystem) {
+				return fmt.Errorf("changing from defined to undefined filesystem is not allowed")
+			}
+			if from.Filesystem != to.Filesystem {
+				return fmt.Errorf("cannot change filesystem from %q to %q",
+					from.Filesystem, to.Filesystem)
+			}
```
Comment (pedronis) [inline gadget/update.go:1505]: same issues here, partial is a permission not to set something but if it is set it still needs to be checked, but maybe I'm missing something

File: gadget/partial.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+			return fmt.Errorf("installer did not provide information for volume %q", volName)
+		}
+
+		// TODO: partial structure, as it is not clear what will be possible when set
+
+		if lov.HasPartial(PartialSchema) {
+			if insVol.Schema == "" {
+				return fmt.Errorf("installer did not provide schema for volume %q", volName)
+			}
+			lov.Schema = insVol.Schema
```
Comment (pedronis) [inline gadget/partial.go:45]: lov.Schema could be already be set?

Comment (pedronis) [reply gadget/partial.go:45]: should we make it an error then to have partial:[schema] but schema set in the volume?

File: gadget/update.go [hunk 8]
```diff
@@ -1474,6 +1496,7 @@ func canUpdateStructure(fromVss []VolumeStructure, fromIdx int, toVss []VolumeSt
 		if !from.HasFilesystem() {
 			return fmt.Errorf("cannot change a bare structure to filesystem one")
 		}
+		// Note that if partial filesystem this will compare empty strings
```
Comment (pedronis) [inline gadget/update.go:1499]: we are not allowing to go from partial no filesystem to filesystem and the other way around? do we need tests for all these combinations?

File: gadget/update.go [hunk 9]
```diff
@@ -1437,21 +1450,30 @@ func arePossibleOffsetsCompatible(vss1 []VolumeStructure, idx1 int, vss2 []Volum
...
 			from.Name, to.Name)
 	}
-	if !arePossibleSizesCompatible(from, to) {
-		return fmt.Errorf("new valid structure size range [%v, %v] is not compatible with current ([%v, %v])",
-			to.MinSize, to.Size, from.MinSize, from.Size)
-	}
-	if !arePossibleOffsetsCompatible(fromVss, fromIdx, toVss, toIdx) {
-		return fmt.Errorf("new valid structure offset range [%v, %v] is not compatible with current ([%v, %v])",
-			minStructureOffset(toVss, toIdx), maxStructureOffset(toVss, toIdx), minStructureOffset(fromVss, fromIdx), maxStructureOffset(fromVss, fromIdx))
+	if !toV.HasPartial(PartialSize) {
```
Comment (pedronis) [inline gadget/update.go:1468]: shouldn't we actually check if the sides actually set a size as well? and the sizes if they are even if in general they are partial and could be omitted

**Review by pedronis — COMMENTED** [review]
some more update.go comments

File: gadget/update.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go:1468]: Right, this is wrong as it is checking the volume, not the structure. Fixed and unit tests added.

File: gadget/partial.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply gadget/partial.go:45]: I have added the check in gadget.go

File: gadget/update.go (hunk 8 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go]: I was thinking on not allowing changes if partial is defined only in one of from/to, but I've changed this now so going from partial to defined is ok - we should allow making the gadget more concrete I think. I've also added unit tests covering the different cases.

File: gadget/update.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go]: Right, this is wrong as it is checking the volume, not the structure. Fixed and unit tests added.

File: tests/lib/muinstaller/main.go [hunk 10]
```diff
@@ -421,6 +418,83 @@ func detectStorageEncryption(seedLabel string) (bool, error) {
...
 }
 
+func fillPartiallyDefinedVolume(vol *gadget.Volume, bootDevice string) error {
+	if len(vol.Partial) == 0 {
+		return nil
+	}
+
+	logger.Noticef("partial gadget for: %q", vol.Partial)
+
+	// TODO: partial structure, as it is not clear what will be possible when set
```
Comment (pedronis) [inline tests/lib/muinstaller/main.go:428]: aren't we testing partial structure at least in so far that there is an extra unknown structure? this TODO maybe needs to be a bit more precise

File: client/systems.go [hunk 11]
```diff
@@ -186,12 +186,21 @@ type SystemDetails struct {
...
 }
 
+func setEnclosingVolumeInStructs(vv map[string]*gadget.Volume) {
+	for _, v := range vv {
+		for sidx := range v.Structure {
+			v.Structure[sidx].EnclosingVolume = v
+		}
+	}
+}
+
```
Comment (pedronis) [inline client/systems.go:196]: feels like this should be a helper in gadget proper? and remove the one in gadgettest?

File: gadget/layout.go [hunk 12]
```diff
@@ -515,7 +515,7 @@ func isLayoutCompatible(current, new *Volume) error {
 
 	// at the structure level we expect the volume to be identical
 	for i := range current.Structure {
-		if err := canUpdateStructure(current.Structure, i, new.Structure, i, new.Schema); err != nil {
+		if err := canUpdateStructure(current, i, new, i); err != nil {
 			return fmt.Errorf("incompatible structure #%d (%q) change: %v", new.Structure[i].YamlIndex, new.Structure[i].Name, err)
 		}
 	}
```
Comment (pedronis) [inline gadget/layout.go:521]: do we need to test partial functionality also from the angle of gadget.IsCompatible? we also still have the question if all these functions can share more code, or we can have fewer functions?

**Review by pedronis — COMMENTED** [review]
thanks, did a full pass again, some comments

File: gadget/layout.go (hunk 12 above)
Comment (alfonsosanchezbeato) [reply gadget/layout.go:520]: I've added some tests using gadget.IsCompatible as entry point. I definitely would like to another gadget refactoring round as a follow-up.

File: tests/lib/muinstaller/main.go (hunk 10 above)
Comment (alfonsosanchezbeato) [reply tests/lib/muinstaller/main.go]: Right, I wouldn't say it is a TODO anymore. I've removed this and instead added a function comment on the things that we fill.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review, comments addressed now. I had to do some additional modifications related to checking Schema properly, which affects some tests, and I removed some unnecessary checks for things that cannot really happen.

File: gadget/gadget.go [hunk 13]
```diff
@@ -255,6 +259,16 @@ type VolumeStructure struct {
 	EnclosingVolume *Volume `yaml:"-" json:"-"`
 }
 
+// SetEnclosingVolumeInStructs is a helper that sets the pointer to
+// the Volume in all VolumeStructure objects it contains.
+func SetEnclosingVolumeInStructs(vv map[string]*Volume) {
```
Comment (pedronis) [inline gadget/gadget.go:264]: this can probably be called just SetEnclosingVolumes, if you think we need the InStructs, then I think it needs to be expanded to the full "InStructures",  as Struct makes one thing of struct types, like reflect.Struct

**Review by pedronis — APPROVED** [review]
thanks, naming comment

File: gadget/partial.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+		if err != nil {
+			return err
+		}
+		if vs.Filesystem == "" && vs.HasFilesystem() {
+			if insStr.Filesystem == "" {
+				return fmt.Errorf("installer did not provide filesystem for structure %q in volume %q", vs.Name, volName)
+			}
+			vs.Filesystem = insStr.Filesystem
+			vsLos.Filesystem = insStr.Filesystem
+		}
```
Comment (Meulengracht) [inline gadget/partial.go:91]: (personal preference) to avoid that level of nesting
```suggestion
		if vs.Filesystem != "" || !vs.HasFilesystem() {
		    continue
		}
                if insStr.Filesystem == "" {
                    return fmt.Errorf("installer did not provide filesystem for structure %q in volume %q", vs.Name, volName)
                }
                vs.Filesystem = insStr.Filesystem
                vsLos.Filesystem = insStr.Filesystem
```

File: gadget/partial.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+			}
+			if insStr.Offset == nil {
+				return fmt.Errorf("installer did not provide offset for structure %q in volume %q", vs.Name, volName)
+			}
+			vs.Size = insStr.Size
+			vsLos.Size = insStr.Size
+			vs.Offset = insStr.Offset
+			vsLos.Offset = insStr.Offset
+			lov.LaidOutStructure[sidx].StartOffset = *insStr.Offset
+		}
```
Comment (Meulengracht) [inline gadget/partial.go:117]: (personal preference) again here it makes sense to reverse the logic checks
```suggestion
		if !vs.hasPartialSize() {
		    continue
		}
	        if insStr.Size == 0 {
		        return fmt.Errorf("installer did not provide size for structure %q in volume %q", vs.Name, volName)
	        }
	        if insStr.Offset == nil {
		        return fmt.Errorf("installer did not provide offset for structure %q in volume %q", vs.Name, volName)
	        }
	        vs.Size = insStr.Size
	        vsLos.Size = insStr.Size
	        vs.Offset = insStr.Offset
	        vsLos.Offset = insStr.Offset
	        lov.LaidOutStructure[sidx].StartOffset = *insStr.Offset
```

File: gadget/update.go [hunk 16]
```diff
@@ -530,23 +532,27 @@ func EnsureVolumeCompatibility(gadgetVolume *Volume, diskVolume *OnDiskVolume, o
 	}
 
 	// Check if top level properties match
```
Comment (Meulengracht) [inline gadget/update.go:534]: nitpick (feel free to ignore): to be honest I'd be great to have a more descriptive comment here, especially now the `if` check is harder to follow, I don't feel like `Check if top level properties match` adds any value in this situation

File: tests/lib/muinstaller/main.go [hunk 17]
```diff
@@ -187,7 +184,7 @@ func createPartitions(bootDevice string, volumes map[string]*gadget.Volume, encT
 	opts := &install.CreateOptions{CreateAllMissingPartitions: true}
 	created, err := install.CreateMissingPartitions(diskLayout, lvol, opts)
 	if err != nil {
-		return nil, fmt.Errorf("cannot create parititons: %v", err)
+		return nil, fmt.Errorf("cannot create partitons: %v", err)
```
Comment (Meulengracht) [inline tests/lib/muinstaller/main.go:187]:
```suggestion
		return nil, fmt.Errorf("cannot create partitions: %v", err)
```

**Review by Meulengracht — APPROVED** [review]
Some minor nitpicks, otherwise LGTM

File: gadget/update.go (hunk 16 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go]: I've changed a bit the comment

File: tests/lib/muinstaller/main.go (hunk 17 above)
Comment (alfonsosanchezbeato) [reply tests/lib/muinstaller/main.go]: Fixed

**Review by alfonsosanchezbeato — COMMENTED** [review]
@Meulengracht thanks for the review, greatly appreciated! I've addressed your comments, and also rebased & cleaned-up commit history.
